### ISSUE TITLE
Exact Match by hashes for Indigo Java API

### ISSUE THREAD

**Issue by AATDev21** [epam/Indigo#775]
We need to switch the existing Java API exact match method to using molecules and reactions hashes to increase its performance.
A similar change was recently implemented in Bingo Elastic.

Here should be noted a special case of multicomponential targets in which an exact match must occur. A few examples for such targets are shown below:    

<img width="713" alt="image" src="https://user-images.githubusercontent.com/78915761/178293468-3325de46-7fe9-498c-8cf4-71028fdf7b4f.png">

**Comment by MysterionRise** [comment]
@AATDev21 makes sense to do it as well. Not sure if it will be faster in Elastic though, did you do performance testing for hash implementation?

Anyway, will add this functionality. Is type of molecules/reactions correct? E.g.

```
MOL_TYPES = ["#02: <molecule>", "#03: <query reaction>", "#12: <RDFMolecule>"]
REAC_TYPES = ["#04: <reaction>", "#05: <query reaction>"]
```

**Comment by AATDev21** [comment]
@MysterionRise, no I didn't.

As for the types of molecules, they definitely were correct at the moment of this issue creation, and I hope they are still ok.
